grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -352,4 +360,75 @@ class PrepaymentInvoiceService {
+        }
+        if (!invoiceItem?.orderAdjustment) {
+            String defaultMessage = "Invoice item is missing order adjustment. Cannot edit unit price."
+            throw new IllegalArgumentException(g.message(code: "invoiceItem.missingAdjustment.error", default: defaultMessage))
+        }
+        BigDecimal totalAdjustments = invoiceItem.orderAdjustment?.totalAdjustments
+        if ((totalAdjustments < 0 && unitPrice > 0) || (totalAdjustments > 0 && unitPrice < 0)) {
+            String defaultMessage = "Cannot change the positive unit price to negative or negative to positive."
+            throw new IllegalArgumentException(g.message(code: "invoiceItem.unitPriceSignChange.error", default: defaultMessage))
+        }
+
+        BigDecimal amountAvailableToInvoice = invoiceItem.orderAdjustment.amountAvailableToInvoice
+        if (Math.abs(unitPrice) > Math.abs(amountAvailableToInvoice) + Math.abs(invoiceItem.unitPrice)) {
+            String defaultMessage = "Cannot update unit price to higher value than available to invoice"
+            throw new IllegalArgumentException(g.message(code: "invoiceItem.unitPriceTooHigh.error", default: defaultMessage))
+        }
+        // update invoice item's unit price (and adjust amount)
+        invoiceItem.unitPrice = unitPrice
+        invoiceItem.amount = invoiceItem.quantity * unitPrice
+        // update inverse item's quantity if there is inverse item (and adjust amount)
+        InvoiceItem inverseItem = findInverseItem(invoiceItem)
+        if (inverseItem) {
+            OrderAdjustment orderAdjustment = invoiceItem?.orderAdjustment
+            InvoiceItem prepaymentItem = orderAdjustment.invoiceItems.find { it.isPrepaymentInvoice }
+            // find quantity that is still available to inverse and add to the current quantity from this inverse item
+            // this needs to be checked because we can also increase invoiced quantity here
+            BigDecimal amountAvailableToInverse = inverseItem.amount + getAmountAvailableToInverse(orderAdjustment, prepaymentItem)
+            BigDecimal amountToInverse = getAmountToInverse(unitPrice, amountAvailableToInverse)
+            inverseItem.unitPrice = amountToInverse
+            inverseItem.amount = inverseItem.quantity * amountToInverse * orderAdjustment.order.prepaymentPercent * (-1)
+            return
+        }
+
+        // If there is a case that we did not found inverse item, but it was made available to inverse now after
+        // editing or removing invoice items on invoices, lets try to create it here
+        inverseItem = createInverseItemForOrderAdjustment(invoiceItem?.orderAdjustment)
+        if (inverseItem) {
+            invoiceItem.invoice.addToInvoiceItems(inverseItem)
+        }
+    }
+
+    BigDecimal getAmountToInverse(BigDecimal amountInvoiced, BigDecimal amountInverseable) {
```

Comment:
I imagine these are public so that you can unit test them but I think it'd be best in general to make methods private whenever possible and then unit test the entrypoint to the service (like the other tests are doing by saving the invoice in the given block).

Reply:
Yeah, I specifically wanted to have easy access to these two methods for unit testing, due to positive and negative cases being a brain teaser.